libkbfs/tlf_journal.go

```diff
@@ -1219,27 +1271,77 @@ func (j *tlfJournal) getMDRange(
 	return j.mdJournal.getRange(start, stop)
 }
 
-func (j *tlfJournal) putMD(ctx context.Context, rmd *RootMetadata) (
-	MdID, error) {
+func (j *tlfJournal) doPutMD(ctx context.Context, rmd *RootMetadata,
+	irmd ImmutableRootMetadata, perRevMap unflushedPathsPerRevMap) (
+	MdID, bool, error) {
+	// Now take the lock and put the MD, merging in the unflushed
+	// paths while under the lock.
 	j.journalLock.Lock()
 	defer j.journalLock.Unlock()
 	if err := j.checkEnabledLocked(); err != nil {
-		return MdID{}, err
+		return MdID{}, false, err
+	}
+
+	if !j.unflushedPaths.appendToCache(irmd, perRevMap) {
+		return MdID{}, true, nil
 	}
+	// TODO: remove the revision from the cache on any errors below?
+	// Tricky when the append is only queued.
 
 	mdID, err := j.mdJournal.put(ctx, j.config.Crypto(),
 		j.config.encryptionKeyGetter(), j.config.BlockSplitter(), rmd)
 	if err != nil {
-		return MdID{}, err
+		return MdID{}, false, err
 	}
 
 	err = j.blockJournal.markMDRevision(ctx, rmd.Revision())
 	if err != nil {
-		return MdID{}, err
+		return MdID{}, false, err
 	}
 
 	j.signalWork()
 
+	return mdID, false, nil
+}
+
+func (j *tlfJournal) putMD(ctx context.Context, rmd *RootMetadata) (
+	MdID, error) {
+	// Prepare the paths without holding the lock, as it might need to
+	// take the lock.  Note that the md ID and timestamp don't matter
+	// for the unflushed path cache.  This is a no-op if the unflushed
+	// path cache is uninitialized.  TODO: avoid doing this if we can
+	// somehow be sure the cache won't be initialized by the time we
+	// finish this put.
+	irmd := MakeImmutableRootMetadata(rmd, fakeMdID(1), time.Now())
+	perRevMap, err := j.unflushedPaths.prepUnflushedPaths(
```

Comment:
No, if something calls `GetJournalStatusAndRange`, the `jStatus.RevisionStart` and `jStatus.RevisionEnd` need to match the unflushed paths returned by that function.